deliver/lib/deliver/upload_screenshots.rb

```diff
@@ -94,5 +130,19 @@ def collect_screenshots_for_languages(path)
 
       return screenshots
     end
+
+    private
+
+    def local_and_remote_match_to_delete(current, checksum)
+      @checksums_local[current.language][current.device_type] &&
+      @checksums_local[current.language][current.device_type].include?(checksum) &&
+      @checksums_local[current.language][current.device_type].index(checksum) + 1 == current.sort_order
+    end
+
+    def local_and_remote_match_to_skip(screenshot, device_type, index, md5)
+      @checksums_remote[screenshot.language][device_type] &&
```

Comment:
I would have introduced the following class:

```
# used to maintain 2 set of MD5s one for local one for remotes
class ScreenshotMD5s
  def initialize
    @checksums = {}
  end

  # index is implicitly computed
  def add_md5(language, device_type, md5)
    @checksums[language] ||= {}
    @checksums_local[language][device_type] ||= []
    @checksums_local[language][device_type] << md5
  end

  def matches_md5(language, device_type, md5, index)
    @checksums[language][device_type] &&
      @checksums[language][device_type].include?(checksum) &&
      @checksums[language][device_type].index(checksum) + 1 == index
   end
end
```

and refactor the whole thing.